pkg/mux/mux.go

```diff
@@ -0,0 +1,272 @@
+package mux
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/bitfinexcom/bitfinex-api-go/pkg/models/event"
+	"github.com/bitfinexcom/bitfinex-api-go/pkg/mux/client"
+	"github.com/bitfinexcom/bitfinex-api-go/pkg/mux/msg"
+)
+
+// Mux will manage all connections and subscriptions. Will check if subscriptions
+// limit is reached and spawn new connection when that happens. It will also listen
+// to all incomming client messages and reconnect client with all its subscriptions
+// in case of a failure
+type Mux struct {
+	cid           int
+	dms           int
+	publicChan    chan msg.Msg
+	publicClients map[int]*client.Client
+	privateChan   chan msg.Msg
+	privateClient *client.Client
+	mtx           *sync.RWMutex
+	Err           error
+	transform     bool
+	apikey        string
+	apisec        string
+	subInfo       map[int64]event.Info
+	authenticated bool
+	publicURL     string
+	authURL       string
+}
+
+// New returns pointer to instance of mux
+func New() *Mux {
+	return &Mux{
+		publicChan:    make(chan msg.Msg),
+		privateChan:   make(chan msg.Msg),
+		publicClients: make(map[int]*client.Client),
+		mtx:           &sync.RWMutex{},
+		subInfo:       map[int64]event.Info{},
+		publicURL:     "wss://api-pub.bitfinex.com/ws/2",
+		authURL:       "wss://api.bitfinex.com/ws/2",
```

Comment:
should be configurable?

Reply:
`func (m *Mux) WithAuthURL(url string) *Mux {`

Reply:
ah it is, via setter